tasks/indexing/task_pdp_ipni.go

```diff
@@ -0,0 +1,427 @@
+	"github.com/filecoin-project/curio/harmony/harmonytask"
+	"github.com/filecoin-project/curio/harmony/resources"
+	"github.com/filecoin-project/curio/harmony/taskhelp"
+	"github.com/filecoin-project/curio/lib/cachedreader"
+	"github.com/filecoin-project/curio/lib/ffi"
+	"github.com/filecoin-project/curio/lib/passcall"
+	"github.com/filecoin-project/curio/market/ipni/chunker"
+	"github.com/filecoin-project/curio/market/ipni/ipniculib"
+)
+
+const (
+	PDP_SP_ID = -2 // This is the SP ID for PDP in the IPNI table
+)
+
+type PDPIPNITask struct {
+	db  *harmonydb.DB
+	cpr *cachedreader.CachedPieceReader
+	cfg *config.CurioConfig
+	max taskhelp.Limiter
+}
+
+func NewPDPIPNITask(db *harmonydb.DB, sc *ffi.SealCalls, cpr *cachedreader.CachedPieceReader, cfg *config.CurioConfig, max taskhelp.Limiter) *PDPIPNITask {
+	return &PDPIPNITask{
+		db:  db,
+		cpr: cpr,
+		cfg: cfg,
+		max: max,
+	}
+}
+
+func (P *PDPIPNITask) Do(taskID harmonytask.TaskID, stillOwned func() bool) (done bool, err error) {
+	ctx := context.Background()
+
+	var tasks []struct {
+		ID       int64  `db:"id"`
+		PieceCID string `db:"piece_cid"`
+		Size     int64  `db:"piece_padded_size"`
+		Prov     string `db:"peer_id"`
+	}
+
+	err = P.db.Select(ctx, &tasks, `
+		SELECT 
+			pr.id,
+			pr.piece_cid,
+			pp.piece_padded_size,
+			ipni_peer.peer_id
+		FROM 
+			pdp_piecerefs pr
+		JOIN parked_piece_refs ppr ON pr.piece_ref = ppr.ref_id
+		JOIN parked_pieces pp ON ppr.piece_id = pp.id
+		CROSS JOIN ipni_peerid ipni_peer
+		WHERE 
+			pr.ipni_task_id = $1
+			AND ipni_peer.sp_id = $2
+`, taskID, PDP_SP_ID)
+	if err != nil {
+		return false, xerrors.Errorf("getting ipni task params: %w", err)
+	}
+
+	if len(tasks) != 1 {
+		return false, xerrors.Errorf("expected 1 ipni task params, got %d", len(tasks))
+	}
+
+	task := tasks[0]
+
+	pcid, err := cid.Parse(task.PieceCID)
+	if err != nil {
+		return false, xerrors.Errorf("parsing piece CID: %w", err)
+	}
+
+	mds := metadata.IpfsGatewayHttp{}
+	md, err := mds.MarshalBinary()
+	if err != nil {
+		return false, xerrors.Errorf("marshaling metadata: %w", err)
+	}
+
+	pi := &PdpIpniContext{
+		PieceCID: pcid,
+		Payload:  true,
+	}
+
+	iContext, err := pi.Marshal()
+	if err != nil {
+		return false, xerrors.Errorf("marshaling piece info: %w", err)
+	}
+
+	reader, _, err := P.cpr.GetSharedPieceReader(ctx, pcid)
+	if err != nil {
+		return false, xerrors.Errorf("getting piece reader: %w", err)
+	}
+	defer func() {
+		_ = reader.Close()
+	}()
+
+	opts := []carv2.Option{carv2.ZeroLengthSectionAsEOF(true)}
+	blockReader, err := carv2.NewBlockReader(bufio.NewReaderSize(reader, 4<<20), opts...)
+	if err != nil {
+		return false, fmt.Errorf("getting block reader over piece: %w", err)
+	}
+
+	chk := chunker.NewInitialChunker()
+
+	blockMetadata, err := blockReader.SkipNext()
+	for err == nil {
+		if err := chk.Accept(blockMetadata.Hash(), int64(blockMetadata.SourceOffset), blockMetadata.Size+40); err != nil {
```

Comment:
```suggestion
		// CAR sections are [varint (length), CID, blockData]
		combinedSize := blockMetadata.Size+c.ByteLen()
		lenSize := varint.UvarintSize(combinedSize)
		sectionSize := combinedSize + lenSize
		if err := chk.Accept(blockMetadata.Hash(), int64(blockMetadata.SourceOffset), sectionSize); err != nil {
```

Looking at the implementation I think this matters: https://github.com/filecoin-project/curio/blob/547a713b148c05ae319ad5dbb7700afdbb0f286c/market/ipni/chunker/initial-chunker.go#L81, although maybe not critical because it's doing chunk size processing? Not hard to be accurate here though.

Hard-coding `40` assumes a certain length varint at the start, followed by a certain length CID, both of which are actually variable.

Also @LexLuthr this is the other spot I was mentioning in https://github.com/filecoin-project/curio/pull/681#issuecomment-3379509599 where you still have `blockMetadata.Offset` in tasks/indexing/task_ipni.go, it should be `SourceOffset` because the chunker seems to care about the first offset in particular.

Reply:
https://github.com/filecoin-project/curio/blob/547a713b148c05ae319ad5dbb7700afdbb0f286c/tasks/indexing/task_ipni.go#L312 is where we take the data in ipni task in main. We are reading a channel here. This channel is fed by https://github.com/filecoin-project/curio/blob/547a713b148c05ae319ad5dbb7700afdbb0f286c/tasks/indexing/task_indexing.go#L438 which uses [sourceOffset](https://github.com/filecoin-project/curio/blob/547a713b148c05ae319ad5dbb7700afdbb0f286c/tasks/indexing/task_indexing.go#L461)

Reply:
In this PR, ipni task is using separate reading mechanism so it needs to be fixed in both places. In main there is only 1 function to read the actual indexes.

Reply:
oh, sorry, I wasn't looking on `main`, I saw the duplicates on `pdpv0` but you've already squashed it down to one place; my bad but I now understand what you were talking about over there

other issue about `sectionSize` vs `40` still stands here, maybe in `main` too